### PR TITLE
experssion: make compare string respect to PAD SPACE

### PR DESCRIPTION
**Pull request by bobotu**
### What problem does this PR solve? 
[MySQL will trim trailing space when comparing nonbinary strings](https://dev.mysql.com/doc/refman/8.0/en/char.html). This PR make TiDB compatible with MySQL.

### What is changed and how it works?
Trim trailing space when comparing nonbinary strings.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

### CODE REVIEW

File: expression/builtin_compare.go [hunk 1]
```diff
@@ -2429,7 +2431,18 @@ func CompareString(sctx sessionctx.Context, lhsArg, rhsArg Expression, lhsRow, r
 	if isNull0 || isNull1 {
 		return compareNull(isNull0, isNull1), true, nil
 	}
-	return int64(types.CompareString(arg0, arg1)), false, nil
+	ret := compareStringWithTrim(arg0, lhsArg.GetType(), arg1, rhsArg.GetType())
+	return int64(ret), false, nil
+}
+
+func compareStringWithTrim(lhs string, lhsFt *types.FieldType, rhs string, rhsFt *types.FieldType) int {
+	if lhsFt.Collate != charset.CollationBin {
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go]: We need some comments here according to to MySQL manual to describe why we need to trim here.

File: expression/builtin_compare.go [hunk 2]
```diff
@@ -2429,7 +2431,18 @@ func CompareString(sctx sessionctx.Context, lhsArg, rhsArg Expression, lhsRow, r
 	if isNull0 || isNull1 {
 		return compareNull(isNull0, isNull1), true, nil
 	}
-	return int64(types.CompareString(arg0, arg1)), false, nil
+	ret := compareStringWithTrim(arg0, lhsArg.GetType(), arg1, rhsArg.GetType())
+	return int64(ret), false, nil
+}
+
+func compareStringWithTrim(lhs string, lhsFt *types.FieldType, rhs string, rhsFt *types.FieldType) int {
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go]: lhs, rhs string, lhsFt, rhsFt *types.FieldTypes

File: types/const_test.go [hunk 3]
```diff
@@ -279,12 +279,12 @@ func (s *testMySQLConstSuite) TestPadCharToFullLengthMode(c *C) {
...
 	r = tk.MustQuery(`SELECT a='xy        ', char_length(a) FROM t1;`)
-	r.Check(testkit.Rows("0 2"))
+	r.Check(testkit.Rows("1 2"))
 	r = tk.MustQuery(`SELECT count(*) FROM t1 WHERE a='xy        ';`)
-	r.Check(testkit.Rows("0"))
+	r.Check(testkit.Rows("1"))
 	tk.MustExec("set sql_mode='PAD_CHAR_TO_FULL_LENGTH';")
 	r = tk.MustQuery(`SELECT a='xy        ', char_length(a) FROM t1;`)
-	r.Check(testkit.Rows("0 2"))
+	r.Check(testkit.Rows("1 2"))
```
Comment (XuHuaiyu) [inline types/const_test.go]: Actually, MySQL returns `1 10` here.
We may add a comment as `FIXME` here.

Comment (Reminiscent) [reply types/const_test.go]: In my mysql 5.7.28
```
mysql> create table t1 (a varchar(10));
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t1 values ('xy');
Query OK, 1 row affected (0.01 sec)

mysql> set sql_mode='PAD_CHAR_TO_FULL_LENGTH';
Query OK, 0 rows affected (0.00 sec)

mysql> SELECT a='xy        ', char_length(a) FROM t1;
+----------------+----------------+
| a='xy        ' | char_length(a) |
+----------------+----------------+
|              1 |              2 |
+----------------+----------------+
1 row in set (0.00 sec)
```

Comment (XuHuaiyu) [reply types/const_test.go]: my fault here

File: types/const_test.go [hunk 4]
```diff
@@ -284,6 +284,7 @@ func (s *testMySQLConstSuite) TestPadCharToFullLengthMode(c *C) {
 	r.Check(testkit.Rows("1"))
 	tk.MustExec("set sql_mode='PAD_CHAR_TO_FULL_LENGTH';")
 	r = tk.MustQuery(`SELECT a='xy        ', char_length(a) FROM t1;`)
+	// FIXME: MySQL returns 1 10 here.
```
Comment (XuHuaiyu) [inline types/const_test.go]: My fault here, this line can be removed.

**Comment by bobotu** [comment]
@bb7133 Insert/Update has been able to deal with this problem correctly, see https://github.com/pingcap/tidb/blob/master/table/column.go#L184-L186

File: expression/builtin_compare.go [hunk 5]
```diff
@@ -2429,7 +2431,20 @@ func CompareString(sctx sessionctx.Context, lhsArg, rhsArg Expression, lhsRow, r
...
 		return compareNull(isNull0, isNull1), true, nil
 	}
-	return int64(types.CompareString(arg0, arg1)), false, nil
+	ret := compareStringWithTrim(arg0, arg1, lhsArg.GetType(), rhsArg.GetType())
+	return int64(ret), false, nil
+}
+
+// According to https://dev.mysql.com/doc/refman/8.0/en/char.html,
+// we need to trim trialing space when compare non-binary strings.
+func compareStringWithTrim(lhs string, rhs string, lhsFt *types.FieldType, rhsFt *types.FieldType) int {
```
Comment (bb7133) [inline expression/builtin_compare.go]: Seems we should consider trimming for both side together, I'm not sure the detailed logic of MySQL, but here is a case I found:

```
(MySQL 8.0)
mysql> select "a " = binary("a ");
+---------------------+
| "a " = binary("a ") |
+---------------------+
|                   1 |
+---------------------+
1 row in set (0.01 sec)

mysql> select "a" = binary("a ");
+--------------------+
| "a" = binary("a ") |
+--------------------+
|                  0 |
+--------------------+
1 row in set (0.00 sec)

(TiDB with this PR)
tidb> select "a " = binary("a ");
+---------------------+
| "a " = binary("a ") |
+---------------------+
|                   0 |
+---------------------+
1 row in set (0.00 sec)

tidb> select "a" = binary("a");
+-------------------+
| "a" = binary("a") |
+-------------------+
|                 1 |
+-------------------+
1 row in set (0.00 sec)

tidb> select "a" = binary("a ");
+--------------------+
| "a" = binary("a ") |
+--------------------+
|                  0 |
+--------------------+
1 row in set (0.00 sec)

(TiDB without this PR)
+---------------------+
| "a " = binary("a ") |
+---------------------+
|                   1 |
+---------------------+
1 row in set (0.01 sec)

tidb> select "a" = binary("a ");
+--------------------+
| "a" = binary("a ") |
+--------------------+
|                  0 |
+--------------------+
1 row in set (0.00 sec)

tidb> select "a " = binary("a");
+--------------------+
| "a " = binary("a") |
+--------------------+
|                  0 |
+--------------------+
1 row in set (0.01 sec)

tidb> select "a" = binary("a");
+-------------------+
| "a" = binary("a") |
+-------------------+
|                 1 |
+-------------------+
1 row in set (0.00 sec)
```

**Comment by zz-jason** [comment]
@winoros @bobotu do we still need to work on this PR since we have abandoned sql mode `pad_char_to_full_length`?

**Comment by zz-jason** [comment]
BTW, this PR is conflicted with the master branch.

**Comment by bobotu** [comment]
@zz-jason I think we can close this PR. There are some thought of this problem https://github.com/pingcap/tidb/pull/13619#issuecomment-568395170.

**Pull request closed without merging**